### PR TITLE
Python Dependencies

### PR DESCRIPTION
**Pull request by joshuatcasey**
Remove the Paketo-hosted dependencies for `poetry` and `pipenv`.

[readable](https://github.com/joshuatcasey/rfcs/blob/jtc/python-dependencies/text/python/0000-dependencies-update.md)

### CODE REVIEW

File: text/python/0000-dependencies-update.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+* poetry: [#known-issues-and-limitations](https://github.com/paketo-buildpacks/poetry#known-issues-and-limitations)
+
+## Rationale
+
+### CPython
+
+Keep this as a Paketo-hosted dependency.
+
+Currently this uses `dep-server`, `binary-builder`, and `buildpacks-ci` to build, but language family maintainers
+will transition this to the new Github Action workflow once that has been approved.
```
Comment (sophiewigmore) [inline text/python/0000-dependencies-update.md:30]: The rationale behind why we need to continue hosting this ourselves is not clear. What's the problem with using the version directly from upstream?

Comment (sophiewigmore) [reply text/python/0000-dependencies-update.md:30]: What's going on in the compilation code that's needed?

File: text/python/0000-dependencies-update.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+As of 2022-06-28, this buildpack is actually not using the dependency from its `buildpack.toml`.
+Instead, the buildpack is running `pip download poetry==$VERSION`, see its [source code](https://github.com/paketo-buildpacks/poetry/blob/ca9e73b34d7018ce00f7d75c76382a08df554a41/poetry_install_process.go#L35).
+
+This RFC does not require Python maintainers to incorporate all logic from `buildpacks-ci` into the `poetry` buildpack.
+
+Also, see [Offline not supported](#offline-not-supported).
+
+#### Poetry layer caching and reproducibility
+
+Remove support for `poetry` layer caching and image reproducibility.
```
Comment (sophiewigmore) [inline text/python/0000-dependencies-update.md:48]: This seems like a pretty hallmark feature of buildpacks. Is there no way to consume Poetry another way? Or to prevent the `pip download` to be skipped on rebuilds?

File: text/python/0000-dependencies-update.md (hunk 1 above)
Comment (robdimsdale) [reply text/python/0000-dependencies-update.md:30]: There are currently [quite a lot of steps](https://github.com/cloudfoundry/buildpacks-ci/blob/ff0027d31d6749d67670b83a7b2da500d992c435/tasks/build-binary-new/builder.rb#L248) to build cpython. We could explore moving all these steps to the buildpack but I don't think it's a trivial change.

File: text/python/0000-dependencies-update.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+Since we rely on `pip` to install `poetry`, the buildpack may not know the version of `poetry` being installed until after the installation completes,
+meaning it is not generally possible to determine if a pre-existing layer can be re-used.
+
+Also, since `poetry` depends on many other packages which can update at any time,
```
Comment (robdimsdale) [inline text/python/0000-dependencies-update.md:53]: I think it is ok to not update transitive dependencies when we have a cached layer. We don't have any reason to believe that this is an issue that consumers face, and the resolution is fairly simple - clear the image cache (e.g. `docker system prune`) and re-run the build.

**Comment by fg-j** [comment]
From Working Group 7/19/22:

- Next steps: RFC will be updated to incorporate a way to implement layer caching for dependencies that are sourced directly from upstream.

**Comment by joshuatcasey** [comment]
TLDR on the changes:

- emphasize that `cpython` will be kept (for now) as a Paketo-hosted dependency due to its complexity
- have deterministic installs of `poetry` and `pipenv` by using `metadata.default-versions`
- ensure that the current layer reuse strategy is preserved with these changes

File: text/python/0000-dependencies-update.md (hunk 1 above)
Comment (joshuatcasey) [reply text/python/0000-dependencies-update.md]: The compilation process involves installing system packages, so this RFC proposes keeping `CPython` as a Paketo-hosted dependency built outside of the buildpack.

File: text/python/0000-dependencies-update.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
+# Decide which Python dependencies will be Paketo-hosted
+
+## Proposal
+
+Currently four Python buildpacks have Paketo-hosted dependencies.
+Only two Paketo-hosted dependencies should be kept, and two should be removed.
```
[the remark this answers is not in the document]
Comment (joshuatcasey) [reply text/python/0000-dependencies-update.md]: I've updated the language

File: text/python/0000-dependencies-update.md (hunk 2 above)
Comment (joshuatcasey) [reply text/python/0000-dependencies-update.md]: I've updated the language to reflect that we want to know (via the `buildpack.yaml[metadata.default-versions]`) what version of the dependency will be installed, and that the dependency layer will be preserved for reuse.

File: text/python/0000-dependencies-update.md (hunk 3 above)
Comment (joshuatcasey) [reply text/python/0000-dependencies-update.md]: Sounds good to me.

**Review by robdimsdale — APPROVED** [review]

**Review by thitch97 — APPROVED** [review]
